From c8dddbd4c935ce534d04be37054a5129036f65b8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Mos=C3=A8=20Giordano?= <mose@gnu.org>
Date: Fri, 1 Aug 2025 09:39:32 +0100
Subject: [PATCH] Update `half` to v2.2.1

This fixes the following compilation error:
```
  >> 153    bazel-out/k8-opt/bin/external/net_sourceforge_half/_virtual_includes/half/half.hpp:911:62: error: '_MM_FROUND_TO_NEAREST_INT' was not declared in this scope
     154        911 |                                 (R==std::round_to_nearest) ? _MM_FROUND_TO_NEAREST_INT :
     155            |                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~
  >> 156      bazel-out/k8-opt/bin/external/net_sourceforge_half/_virtual_includes/half/half.hpp:912:63: error: '_MM_FROUND_TO_ZERO' was not declared in this scope
     157        912 |                                 (R==std::round_toward_zero) ? _MM_FROUND_TO_ZERO :
     158            |                                                               ^~~~~~~~~~~~~~~~~~
```

Taken out of b67e9581a1a9d516a0be3efe0ed00218e7cfe250.
---
 .../patches/detail_raise.patch                | 157 ------------------
 .../net_sourceforge_half/workspace.bzl        |  15 +-
 2 files changed, 6 insertions(+), 166 deletions(-)
 delete mode 100644 third_party/net_sourceforge_half/patches/detail_raise.patch

diff --git a/third_party/net_sourceforge_half/patches/detail_raise.patch b/third_party/net_sourceforge_half/patches/detail_raise.patch
deleted file mode 100644
index e5382315..00000000
--- a/third_party/net_sourceforge_half/patches/detail_raise.patch
+++ /dev/null
@@ -1,157 +0,0 @@
---- a/include/half.hpp
-+++ b/include/half.hpp
-@@ -661,11 +661,11 @@
- 		#endif
- 		#if HALF_ERRHANDLING_UNDERFLOW_TO_INEXACT
- 			if((flags & FE_UNDERFLOW) && !(flags & FE_INEXACT))
--				raise(FE_INEXACT);
-+				detail::raise(FE_INEXACT);
- 		#endif
- 		#if HALF_ERRHANDLING_OVERFLOW_TO_INEXACT
- 			if((flags & FE_OVERFLOW) && !(flags & FE_INEXACT))
--				raise(FE_INEXACT);
-+				detail::raise(FE_INEXACT);
- 		#endif
- 		#endif
- 		}
-@@ -679,7 +679,7 @@
- 		inline HALF_CONSTEXPR_NOERR bool compsignal(unsigned int x, unsigned int y)
- 		{
- 		#if HALF_ERRHANDLING
--			raise(FE_INVALID, (x&0x7FFF)>0x7C00 || (y&0x7FFF)>0x7C00);
-+			detail::raise(FE_INVALID, (x&0x7FFF)>0x7C00 || (y&0x7FFF)>0x7C00);
- 		#endif
- 			return (x&0x7FFF) > 0x7C00 || (y&0x7FFF) > 0x7C00;
- 		}
-@@ -691,7 +691,7 @@
- 		inline HALF_CONSTEXPR_NOERR unsigned int signal(unsigned int nan)
- 		{
- 		#if HALF_ERRHANDLING
--			raise(FE_INVALID, !(nan&0x200));
-+			detail::raise(FE_INVALID, !(nan&0x200));
- 		#endif
- 			return nan | 0x200;
- 		}
-@@ -704,7 +704,7 @@
- 		inline HALF_CONSTEXPR_NOERR unsigned int signal(unsigned int x, unsigned int y)
- 		{
- 		#if HALF_ERRHANDLING
--			raise(FE_INVALID, ((x&0x7FFF)>0x7C00 && !(x&0x200)) || ((y&0x7FFF)>0x7C00 && !(y&0x200)));
-+			detail::raise(FE_INVALID, ((x&0x7FFF)>0x7C00 && !(x&0x200)) || ((y&0x7FFF)>0x7C00 && !(y&0x200)));
- 		#endif
- 			return ((x&0x7FFF)>0x7C00) ? (x|0x200) : (y|0x200);
- 		}
-@@ -718,7 +718,7 @@
- 		inline HALF_CONSTEXPR_NOERR unsigned int signal(unsigned int x, unsigned int y, unsigned int z)
- 		{
- 		#if HALF_ERRHANDLING
--			raise(FE_INVALID, ((x&0x7FFF)>0x7C00 && !(x&0x200)) || ((y&0x7FFF)>0x7C00 && !(y&0x200)) || ((z&0x7FFF)>0x7C00 && !(z&0x200)));
-+			detail::raise(FE_INVALID, ((x&0x7FFF)>0x7C00 && !(x&0x200)) || ((y&0x7FFF)>0x7C00 && !(y&0x200)) || ((z&0x7FFF)>0x7C00 && !(z&0x200)));
- 		#endif
- 			return ((x&0x7FFF)>0x7C00) ? (x|0x200) : ((y&0x7FFF)>0x7C00) ? (y|0x200) : (z|0x200);
- 		}
-@@ -743,7 +743,7 @@
- 		inline HALF_CONSTEXPR_NOERR unsigned int invalid()
- 		{
- 		#if HALF_ERRHANDLING
--			raise(FE_INVALID);
-+			detail::raise(FE_INVALID);
- 		#endif
- 			return 0x7FFF;
- 		}
-@@ -755,7 +755,7 @@
- 		inline HALF_CONSTEXPR_NOERR unsigned int pole(unsigned int sign = 0)
- 		{
- 		#if HALF_ERRHANDLING
--			raise(FE_DIVBYZERO);
-+			detail::raise(FE_DIVBYZERO);
- 		#endif
- 			return sign | 0x7C00;
- 		}
-@@ -767,7 +767,7 @@
- 		inline HALF_CONSTEXPR_NOERR unsigned int check_underflow(unsigned int arg)
- 		{
- 		#if HALF_ERRHANDLING && !HALF_ERRHANDLING_UNDERFLOW_TO_INEXACT
--			raise(FE_UNDERFLOW, !(arg&0x7C00));
-+			detail::raise(FE_UNDERFLOW, !(arg&0x7C00));
- 		#endif
- 			return arg;
- 		}
-@@ -784,7 +784,7 @@
- 		template<std::float_round_style R> HALF_CONSTEXPR_NOERR unsigned int overflow(unsigned int sign = 0)
- 		{
- 		#if HALF_ERRHANDLING
--			raise(FE_OVERFLOW);
-+			detail::raise(FE_OVERFLOW);
- 		#endif
- 			return	(R==std::round_toward_infinity) ? (sign+0x7C00-(sign>>15)) :
- 					(R==std::round_toward_neg_infinity) ? (sign+0x7BFF+(sign>>15)) :
-@@ -800,7 +800,7 @@
- 		template<std::float_round_style R> HALF_CONSTEXPR_NOERR unsigned int underflow(unsigned int sign = 0)
- 		{
- 		#if HALF_ERRHANDLING
--			raise(FE_UNDERFLOW);
-+			detail::raise(FE_UNDERFLOW);
- 		#endif
- 			return	(R==std::round_toward_infinity) ? (sign+1-(sign>>15)) :
- 					(R==std::round_toward_neg_infinity) ? (sign+(sign>>15)) :
-@@ -824,11 +824,11 @@
- 						(R==std::round_toward_infinity) ? (~(value>>15)&(g|s)) :
- 						(R==std::round_toward_neg_infinity) ? ((value>>15)&(g|s)) : 0;
- 			if((value&0x7C00) == 0x7C00)
--				raise(FE_OVERFLOW);
-+				detail::raise(FE_OVERFLOW);
- 			else if(value & 0x7C00)
--				raise(FE_INEXACT, I || (g|s)!=0);
-+				detail::raise(FE_INEXACT, I || (g|s)!=0);
- 			else
--				raise(FE_UNDERFLOW, !(HALF_ERRHANDLING_UNDERFLOW_TO_INEXACT) || I || (g|s)!=0);
-+				detail::raise(FE_UNDERFLOW, !(HALF_ERRHANDLING_UNDERFLOW_TO_INEXACT) || I || (g|s)!=0);
- 			return value;
- 		#else
- 			return	(R==std::round_to_nearest) ? (value+(g&(s|value))) :
-@@ -851,7 +851,7 @@
- 			unsigned int abs = value & 0x7FFF;
- 			if(abs < 0x3C00)
- 			{
--				raise(FE_INEXACT, I);
-+				detail::raise(FE_INEXACT, I);
- 				return ((R==std::round_to_nearest) ? (0x3C00&-static_cast<unsigned>(abs>=(0x3800+E))) :
- 						(R==std::round_toward_infinity) ? (0x3C00&-(~(value>>15)&(abs!=0))) :
- 						(R==std::round_toward_neg_infinity) ? (0x3C00&-static_cast<unsigned>(value>0x8000)) :
-@@ -860,7 +860,7 @@
- 			if(abs >= 0x6400)
- 				return (abs>0x7C00) ? signal(value) : value;
- 			unsigned int exp = 25 - (abs>>10), mask = (1<<exp) - 1;
--			raise(FE_INEXACT, I && (value&mask));
-+			detail::raise(FE_INEXACT, I && (value&mask));
- 			return ((	(R==std::round_to_nearest) ? ((1<<(exp-1))-(~(value>>exp)&E)) :
- 						(R==std::round_toward_infinity) ? (mask&((value>>15)-1)) :
- 						(R==std::round_toward_neg_infinity) ? (mask&-(value>>15)) :
-@@ -1319,12 +1319,12 @@
- 			unsigned int abs = value & 0x7FFF;
- 			if(abs >= 0x7C00)
- 			{
--				raise(FE_INVALID);
-+				detail::raise(FE_INVALID);
- 				return (value&0x8000) ? std::numeric_limits<T>::min() : std::numeric_limits<T>::max();
- 			}
- 			if(abs < 0x3800)
- 			{
--				raise(FE_INEXACT, I);
-+				detail::raise(FE_INEXACT, I);
- 				return	(R==std::round_toward_infinity) ? T(~(value>>15)&(abs!=0)) :
- 						(R==std::round_toward_neg_infinity) ? -T(value>0x8000) :
- 						T();
-@@ -1337,9 +1337,9 @@
- 				(R==std::round_toward_neg_infinity) ? (((1<<exp)-1)&-(value>>15)) : 0))>>exp));
- 			if((!std::numeric_limits<T>::is_signed && (value&0x8000)) || (std::numeric_limits<T>::digits<16 &&
- 				((value&0x8000) ? (-i<std::numeric_limits<T>::min()) : (i>std::numeric_limits<T>::max()))))
--				raise(FE_INVALID);
-+				detail::raise(FE_INVALID);
- 			else if(I && exp > 0 && (m&((1<<exp)-1)))
--				raise(FE_INEXACT);
-+				detail::raise(FE_INEXACT);
- 			return static_cast<T>((value&0x8000) ? -i : i);
- 		}
- 
diff --git a/third_party/net_sourceforge_half/workspace.bzl b/third_party/net_sourceforge_half/workspace.bzl
index 7ae937e6..ea0a31b3 100644
--- a/third_party/net_sourceforge_half/workspace.bzl
+++ b/third_party/net_sourceforge_half/workspace.bzl
@@ -12,27 +12,24 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# buildifier: disable=module-docstring
+
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "maybe")
 load(
     "//third_party:repo.bzl",
     "third_party_http_archive",
 )
-load("@bazel_tools//tools/build_defs/repo:utils.bzl", "maybe")
 
 def repo():
     maybe(
         third_party_http_archive,
         name = "net_sourceforge_half",
         urls = [
-            "https://storage.googleapis.com/tensorstore-bazel-mirror/sourceforge.net/projects/half/files/half/2.1.0/half-2.1.0.zip",
-            "https://sourceforge.net/projects/half/files/half/2.1.0/half-2.1.0.zip",
+            "https://storage.googleapis.com/tensorstore-bazel-mirror/sourceforge.net/projects/half/files/half/2.2.1/half-2.2.1.zip",
+            "https://sourceforge.net/projects/half/files/half/2.2.1/half-2.2.1.zip",
         ],
-        sha256 = "ad1788afe0300fa2b02b0d1df128d857f021f92ccf7c8bddd07812685fa07a25",
+        sha256 = "76ddbf406e9d9b772ec73af2bf925b38b290b4390cc4064720a08d4b4bca0aa9",
         build_file = Label("//third_party:net_sourceforge_half/half.BUILD.bazel"),
-        patches = [
-            # https://sourceforge.net/p/half/discussion/general/thread/86298c105c/
-            Label("//third_party:net_sourceforge_half/patches/detail_raise.patch"),
-        ],
-        patch_args = ["-p1"],
         cmake_name = "half",
         cmake_target_mapping = {
             "@net_sourceforge_half//:half": "half::half",
-- 
2.31.0

